File: src/common/wasm/unravel_index.ts
/**
 * WASM-accelerated unravel_index: convert flat indices to multi-dimensional indices.
 * Uses native i32/i64 kernels for integer inputs (no JS conversion overhead),
 * with f64 kernel as fallback. All kernels output f64 directly.
 */

import {
  unravel_index_i32_f64,
  unravel_index_i64_f64,
  unravel_index_f64,
} from './bins/unravel_index.wasm';
import {
  wasmMalloc,
  resetScratchAllocator,
  resolveInputPtr,
  scratchAlloc,
  scratchCopyIn,
  getSharedMemory,
} from './runtime';
import type { WasmRegion } from './runtime';
import { ArrayStorage } from '../storage';
import type { TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 32;

/**
 * WASM-accelerated unravel_index for flat indices with C-order strides.
 * Returns ndim ArrayStorage arrays (float64), or null if WASM can't handle this case.
 */
export function wasmUnravelIndex(indices: ArrayStorage, shape: number[]): ArrayStorage[] | null {
  if (!indices.isCContiguous) return null;

  const N = indices.size;
  const ndim = shape.length;
  if (ndim === 0 || N === 0) return null;
  if (N < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = indices.dtype;
  const idxData = indices.data;
  const idxOff = indices.offset;
  const f64Bpe = 8;

  // Allocate f64 output regions
  const outRegions: WasmRegion[] = [];
  for (let d = 0; d < ndim; d++) {
    const region = wasmMalloc(N * f64Bpe);
    if (!region) {
      for (const r of outRegions) r.release();
      return null;
    }
    outRegions.push(region);
  }

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  // All kernels output f64 directly into scratch, then copy to persistent regions
  const outScratchPtr = scratchAlloc(ndim * N * f64Bpe);

  if (dtype === 'int64' || dtype === 'uint64') {
    // i64 kernel: native BigInt input, f64 output
    const bpe64 = 8;
    const indicesPtr = resolveInputPtr(
      idxData,
      indices.isWasmBacked,
      indices.wasmPtr,
      idxOff,
      N,
      bpe64
    );

    const strides64 = new BigInt64Array(ndim);
    let stride64 = 1n;
    for (let i = ndim - 1; i >= 0; i--) {
      strides64[i] = stride64;
      stride64 *= BigInt(shape[i]!);
    }
    const stridesPtr = scratchCopyIn(strides64 as unknown as TypedArray);
    const shape64 = new BigInt64Array(shape.map(BigInt));
    const shapePtr = scratchCopyIn(shape64 as unknown as TypedArray);

    unravel_index_i64_f64(indicesPtr, outScratchPtr, N, stridesPtr, shapePtr, ndim);
  } else if (dtype === 'int32' || dtype === 'uint32') {
    // i32 kernel: native int input, f64 output
    const bpe32 = 4;
    const indicesPtr = resolveInputPtr(
      idxData,
      indices.isWasmBacked,
      indices.wasmPtr,
      idxOff,
      N,
      bpe32
    );

    const strides = new Int32Array(ndim);
    let stride = 1;
    for (let i = ndim - 1; i >= 0; i--) {
      strides[i] = stride;
      stride *= shape[i]!;
    }
    const stridesPtr = scratchCopyIn(strides as unknown as TypedArray);
    const shapeI32 = new Int32Array(shape);
    const shapePtr = scratchCopyIn(shapeI32 as unknown as TypedArray);

    unravel_index_i32_f64(indicesPtr, outScratchPtr, N, stridesPtr, shapePtr, ndim);
  } else {
    // f64 fallback: convert input to f64 if needed
    let indicesPtr: number;
    if (dtype === 'float64') {
      indicesPtr = resolveInputPtr(
        idxData,
        indices.isWasmBacked,
        indices.wasmPtr,
        idxOff,
        N,
        f64Bpe
      );
    } else {
      const f64 = new Float64Array(N);
      for (let i = 0; i < N; i++) f64[i] = Number(idxData[idxOff + i]);
      indicesPtr = scratchCopyIn(f64 as unknown as TypedArray);
    }

    const strides = new Int32Array(ndim);
    let stride = 1;
    for (let i = ndim - 1; i >= 0; i--) {
      strides[i] = stride;
      stride *= shape[i]!;
    }
    const stridesPtr = scratchCopyIn(strides as unknown as TypedArray);
    const shapeI32 = new Int32Array(shape);
    const shapePtr = scratchCopyIn(shapeI32 as unknown as TypedArray);

    unravel_index_f64(indicesPtr, outScratchPtr, N, stridesPtr, shapePtr, ndim);
  }

  // Copy f64 scratch output to persistent regions
  const mem = getSharedMemory();
  for (let d = 0; d < ndim; d++) {
    const src = new Float64Array(mem.buffer, outScratchPtr + d * N * f64Bpe, N);
    const dst = new Float64Array(mem.buffer, outRegions[d]!.ptr, N);
    dst.set(src);
  }

  const outputShape = Array.from(indices.shape);
  const results: ArrayStorage[] = [];
  for (let d = 0; d < ndim; d++) {
    results.push(
      ArrayStorage.fromWasmRegion(
        outputShape,
        'float64',
        outRegions[d]!,
        N,
        Float64Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray
      )
    );
  }

  return results;
}

File: src/common/wasm/zig/unravel_index.zig
//! Unravel index implementation for i32/u32 and i64/u64 flat indices, outputting f64 for JS compatibility.
//! Uses i32/i64 arithmetic internally for exact integer calculations, outputs f64 for JS Number compatibility.
//! Also includes a version for f64 input

/// Unravel flat indices into multi-dimensional indices (i32/u32 input, f64 output).
/// Uses i32 arithmetic internally, outputs f64 for JS Number compatibility.
/// Works for both i32 and u32 indices (u32 reinterprets as i32 — safe since indices are non-negative).
export fn unravel_index_i32_f64(indices: [*]const i32, out: [*]f64, N: u32, strides: [*]const i32, shape: [*]const i32, ndim: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        var remaining: i32 = indices[i];
        var d: u32 = 0;
        while (d < ndim) : (d += 1) {
            const s = strides[d];
            const idx = @divTrunc(remaining, s);
            out[d * N + i] = @floatFromInt(@rem(idx, shape[d]));
            remaining = @rem(remaining, s);
        }
    }
}

/// Unravel flat indices into multi-dimensional indices (i64/u64 input, f64 output).
/// Uses i64 arithmetic internally, outputs f64 for JS Number compatibility.
/// Works for both i64 and u64 indices (indices are always non-negative).
export fn unravel_index_i64_f64(indices: [*]const i64, out: [*]f64, N: u32, strides: [*]const i64, shape: [*]const i64, ndim: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        var remaining: i64 = indices[i];
        var d: u32 = 0;
        while (d < ndim) : (d += 1) {
            const s = strides[d];
            const idx = @divTrunc(remaining, s);
            out[d * N + i] = @floatFromInt(@rem(idx, shape[d]));
            remaining = @rem(remaining, s);
        }
    }
}

/// Unravel flat indices (f64 input, f64 output).
/// For non-integer input types that get converted to f64 on the JS side.
/// Computes in i64 internally for exact integer arithmetic.
export fn unravel_index_f64(indices: [*]const f64, out: [*]f64, N: u32, strides: [*]const i32, shape: [*]const i32, ndim: u32) void {
    var i: u32 = 0;
    while (i < N) : (i += 1) {
        var remaining: i64 = @intFromFloat(indices[i]);
        var d: u32 = 0;
        while (d < ndim) : (d += 1) {
            const s: i64 = strides[d];
            const idx = @divTrunc(remaining, s);
            out[d * N + i] = @floatFromInt(@rem(idx, @as(i64, shape[d])));
            remaining = @rem(remaining, s);
        }
    }
}

// --- Tests ---

test "unravel_index_i32_f64 basic 2D" {
    const testing = @import("std").testing;
    const indices = [_]i32{ 0, 5, 11 };
    const strides = [_]i32{ 4, 1 };
    const shape = [_]i32{ 3, 4 };
    var out: [6]f64 = undefined;
    unravel_index_i32_f64(&indices, &out, 3, &strides, &shape, 2);
    try testing.expectEqual(out[0], 0.0);
    try testing.expectEqual(out[3], 0.0);
    try testing.expectEqual(out[1], 1.0);
    try testing.expectEqual(out[4], 1.0);
    try testing.expectEqual(out[2], 2.0);
    try testing.expectEqual(out[5], 3.0);
}

test "unravel_index_i32_f64 3D" {
    const testing = @import("std").testing;
    const indices = [_]i32{ 0, 23 };
    const strides = [_]i32{ 12, 4, 1 };
    const shape = [_]i32{ 2, 3, 4 };
    var out: [6]f64 = undefined;
    unravel_index_i32_f64(&indices, &out, 2, &strides, &shape, 3);
    try testing.expectEqual(out[0], 0.0);
    try testing.expectEqual(out[2], 0.0);
    try testing.expectEqual(out[4], 0.0);
    try testing.expectEqual(out[1], 1.0);
    try testing.expectEqual(out[3], 2.0);
    try testing.expectEqual(out[5], 3.0);
}

test "unravel_index_i64_f64 basic 2D" {
    const testing = @import("std").testing;
    const indices = [_]i64{ 0, 5, 11 };
    const strides = [_]i64{ 4, 1 };
    const shape = [_]i64{ 3, 4 };
    var out: [6]f64 = undefined;
    unravel_index_i64_f64(&indices, &out, 3, &strides, &shape, 2);
    try testing.expectEqual(out[0], 0.0);
    try testing.expectEqual(out[3], 0.0);
    try testing.expectEqual(out[1], 1.0);
    try testing.expectEqual(out[4], 1.0);
    try testing.expectEqual(out[2], 2.0);
    try testing.expectEqual(out[5], 3.0);
}

test "unravel_index_f64 basic 2D" {
    const testing = @import("std").testing;
    const indices = [_]f64{ 0.0, 5.0, 11.0 };
    const strides = [_]i32{ 4, 1 };
    const shape = [_]i32{ 3, 4 };
    var out: [6]f64 = undefined;
    unravel_index_f64(&indices, &out, 3, &strides, &shape, 2);
    try testing.expectEqual(out[0], 0.0);
    try testing.expectEqual(out[3], 0.0);
    try testing.expectEqual(out[1], 1.0);
    try testing.expectEqual(out[4], 1.0);
    try testing.expectEqual(out[2], 2.0);
    try testing.expectEqual(out[5], 3.0);
}

test "unravel_index_f64 3D" {
    const testing = @import("std").testing;
    const indices = [_]f64{ 0.0, 23.0 };
    const strides = [_]i32{ 12, 4, 1 };
    const shape = [_]i32{ 2, 3, 4 };
    var out: [6]f64 = undefined;
    unravel_index_f64(&indices, &out, 2, &strides, &shape, 3);
    try testing.expectEqual(out[0], 0.0);
    try testing.expectEqual(out[2], 0.0);
    try testing.expectEqual(out[4], 0.0);
    try testing.expectEqual(out[1], 1.0);
    try testing.expectEqual(out[3], 2.0);
    try testing.expectEqual(out[5], 3.0);
}
